backends/p4fmt/p4formatter.cpp

```diff
@@ -0,0 +1,1515 @@
+                    builder.append(sourceFile);
+                    builder.append("\"");
+                    builder.newline();
+                }
+                includesEmitted.emplace(sourceFile);
+            }
+            first = false;
+            continue;
+        }
+        if (!first) builder.newline();
+        first = false;
+        visit(a);
+    }
+    if (!program->objects.empty()) builder.newline();
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Bits *t) {
+    if (t->expression) {
+        builder.append("bit<(");
+        visit(t->expression);
+        builder.append(")>");
+    } else {
+        builder.append(t->toString());
+    }
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_String *t) {
+    builder.append(t->toString());
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_InfInt *t) {
+    builder.append(t->toString());
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Var *t) {
+    builder.append(t->name);
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Unknown *) {
+    BUG("Cannot emit code for an unknown type");
+    // builder.append("*unknown type*");
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Dontcare *) {
+    builder.append("_");
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Void *) {
+    builder.append("void");
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Name *t) {
+    visit(t->path);
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Stack *t) {
+    visit(t->elementType);
+    builder.append("[");
+    visit(t->size);
+    builder.append("]");
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Specialized *t) {
+    visit(t->baseType);
+    builder.append("<");
+    setVecSep(", ");
+    visit(t->arguments);
+    doneVec();
+    builder.append(">");
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Argument *arg) {
+    if (!arg->name.name.isNullOrEmpty()) {
+        builder.append(arg->name.name);
+        builder.append(" = ");
+    }
+    visit(arg->expression);
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Typedef *t) {
+    if (!t->annotations->annotations.empty()) {
+        visit(t->annotations);
+        builder.spc();
+    }
+    builder.append("typedef ");
+    visit(t->type);
+    builder.spc();
+    builder.append(t->name);
+    builder.endOfStatement();
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_Newtype *t) {
+    if (!t->annotations->annotations.empty()) {
+        visit(t->annotations);
+        builder.spc();
+    }
+    builder.append("type ");
+    visit(t->type);
+    builder.spc();
+    builder.append(t->name);
+    builder.endOfStatement();
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_BaseList *t) {
+    builder.append("tuple<");
+    bool first = true;
+    for (auto a : t->components) {
```

Comment:
This looks like a common pattern (visit some collection with comma separator and this `fist` /  not-`first`). Consider refactoring into some common code of visitation with separator. E.g. you can submit a lambda to visit the internals, e.g.:
```c++
visit_collection(t->components, ",", 
                 [](const auto *elem) {
                      auto p4type = a->getP4Type();
                      CHECK_NULL(p4type);
                      visit(p4type);
                   });
```